### ISSUE TITLE
Use SQLAlchemy ORM and avoid runnign SQL queries directly

### ISSUE THREAD

**Issue by kinow** [cylc/cylc#2800]
There are a few warnings in Codacy about security issues regarding concatenating strings when querying the sqlite database.

These queries are SQL statements, and in most cases it is the table name that is being concatenated.

So instead of writing these queries, we could possibly rely on SQLAlchemy and use its ORM resources like Query and Session to use something similar to

```python
# untested code, just to illustrate a possible solution

Base = declarative_base()

class CheckPointId(Base):
    __tablename__ = 'checkpointid'

    id = Column(Integer, primary_key=True)

class CheckPointIdRepository(object):
    def __init__(self, session):
        self.session = session
    def count_checkpoints(self):
        return session.query(func.count(CheckPointId.id)).scalar()
```

The main benefit for SQLAlchemy here is the security and database mapping. Downsides are complexity, and slight performance (though insignificant if you compare with Java's Hibernate).

We could also re-use some of the entities from the JupyterHub project for the user management. Killing two birds here, the annoying security warnings from Codacy, and re-use existing code.

**Comment by matthewrmshin** [comment]
Yes. This has been debated before we added SQLite storage for suite run time data. The direct usage of SQLite has the advantage of performance, and we are not really doing anything complex to require a SQL statement builder.

The main problem Codacy seems to be complaining about is the use of local class constants for table names for constructing the SQL statement or statement template. I cannot really see a route to inject SQL statements this way, but I am not a security expert either. I suppose this is because Python constants are not real constants but just a convention - and there may be routes for them to be changed? Ironically, if we hard code (instead of using the table name string constants), then Codacy would stop complaining. Easy enough to fix for the sake of security, but does not make sense in terms of code sanity.

**Comment by hjoliver** [comment]
@kinow - is there any good reason to do this other than to squash some Codacy warnings?  Specifically, are these warnings actually meaningful - i.e. a possible route to SQL injection - in any of our cases?  If not, then it seems a bit persverse to make our code more complicated to please a tools that is not sufficiently intelligent to detect real problems (as apposed to potential problems)...

**Comment by kinow** [comment]
I agree with @matthewrmshin, that it's quite hard to exploit the parts of code in these warnings (I think I saw one or two places where we were actually attaching contents of messages to the query).

But the sqlite API provides enough security to allow us to escape parameters and avoid SQL injections.

@hjoliver good points. The tool that I am looking at is actually something to put between the SQL and the code in that DAO layer. At least in Java we have an intermediate solution, that allows for query and entity mapping, without the complexity and performance penalty (jOOQ), but alas I am not aware of a Python alternative.

Will close this ticket, and sleep on it for now :+1:

**Comment by matthewrmshin** [comment]
Looks like we should be able to get rid of the issue from `select_task_job`, because `cylc cat-log` calls it with `keys=None`, so we just need to return all columns by `SELECT * ...` - there is no need to inject the string.
